File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/M._Проверка_на_двудольность/solution.py
import sys


def is_bipartite(source: int, adj_list: dict[int, list[int]], colors: list[int]) -> bool:
    call_stack: list[int] = [source]
    while call_stack:
        vertex: int = call_stack.pop()
        for neighbor in adj_list.get(vertex, []):
            if colors[neighbor] == colors[vertex]: return False
            elif colors[neighbor] != -1: continue
            colors[neighbor] = (colors[vertex] + 1) % 2
            call_stack.append(neighbor)
    return True


def solution() -> None:
    """
    Time Complexity O(V + E)
    Memory Complexity O(V + E)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    adj_list: dict[int, list[int]] = dict()
    for _ in range(m):
        u, v = map(int, sys.stdin.readline().rstrip().split())
        if u not in adj_list: adj_list[u] = list()
        adj_list[u].append(v)
        if v not in adj_list: adj_list[v] = list()
        adj_list[v].append(u)
    colors: list[int] = [-1] * (n + 1)
    for vertex in range(1, n + 1):
        if colors[vertex] != -1: continue
        colors[vertex] = 0
        if not is_bipartite(vertex, adj_list, colors):
            sys.stdout.write('NO\n')
            return
    sys.stdout.write('YES\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/M._Проверка_на_двудольность/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

bool isBipartite(int source, std::unordered_map<int, std::vector<int>>& adjList, std::vector<int>& colors) {
    std::vector<int> callStack = {source};
    while (!callStack.empty()) {
        int vertex = callStack.back();
        callStack.pop_back();
        for (int neighbor : adjList[vertex]) {
            if (colors[neighbor] == colors[vertex]) return false;
            else if (colors[neighbor] != -1) continue;
            colors[neighbor] = (colors[vertex] + 1) % 2;
            callStack.push_back(neighbor);
        }
    }
    return true;
}

void solution() {
    /*
    Time Complexity O(V + E)
    Memory Complexity O(V + E)
    */
    int n, m;
    std::cin >> n >> m;
    std::unordered_map<int, std::vector<int>> adjList;
    for (int i = 0; i < m; ++i) {
        int u, v;
        std::cin >> u >> v;
        adjList[u].push_back(v);
        adjList[v].push_back(u);
    }
    std::vector<int> colors (n + 1, -1);
    for (int vertex = 1; vertex <= n; ++vertex) {
        if (colors[vertex] != -1) continue;
        colors[vertex] = 0;
        if (!isBipartite(vertex, adjList, colors)) {
            std::cout << "NO\n";
            return;
        }
    }
    std::cout << "YES\n";
}


int main() {
    solution();
}